include/picongpu/particles/collision/binary/DefaultAlg.hpp

```diff
@@ -0,0 +1,499 @@
+                float_X cosXi = math::log( bracket ) / a;
+                if ( cosXi < -1.0_X || cosXi > 1.0_X )
+                {
+                    if  (cosXi < -1.0_X ) return -1.0_X;
+                    else return 1.0_X;
+                }
+                else return cosXi;
+            }
+            else
+            {
+                // + 1 in Perez 2012 but cos in [-1,1]. Smilei uses -1.
+                return 2 * u - 1;
+            }
+        }
+
+        // (12) in [Perez 2012]
+        static DINLINE float3_X calcFinalComsMomentum(
+            float3_X const p,
+            float_X const cosXi,
+            float_X const phi
+        )
+        {
+            float_X sinPhi, cosPhi;
+            pmacc::math::sincos( phi, sinPhi, cosPhi);
+            float_X sinXi = math::sqrt( 1 - cosXi * cosXi );
+
+            float3_X finalVec;
+            float_X const pAbs = math::sqrt( pmacc::math::abs2( p ) );
+            float_X const pPerp = math::sqrt( p.x( ) * p.x( ) + p.y( )
+                * p.y( ) );
+            // TODO chose a better limmit?
+            if( pPerp > std::max( std::numeric_limits< float_X >::epsilon, 1.0e-10_X ) *  pAbs )
+            {
+                finalVec[ 0 ] = ( p.x( ) * p.z( ) * sinXi * cosPhi
+                    - p.y( ) * pAbs * sinXi * sinPhi ) / pPerp + p.x( ) * cosXi;
+                finalVec[ 1 ] = ( p.y( ) * p.z( ) * sinXi * cosPhi
+                   + p.x( ) * pAbs * sinXi * sinPhi ) / pPerp + p.y( ) * cosXi;
+                finalVec[ 2 ] = -1.0_X * pPerp * sinXi * cosPhi
+                    + p.z( ) * cosXi;
+            }
+            else // limit px->0 py=0
+            {
+                finalVec[ 0 ] = pAbs * sinXi * cosPhi;
+                finalVec[ 1 ] = pAbs * sinXi * sinPhi;
+                finalVec[ 2 ] = pAbs * cosXi;
+            }
+            return finalVec;
+        }
+
+        /** execute the user manipulator functor
+         *
+         * @tparam T_Args type of the arguments passed to the user manipulator
+         *     functor.
+         *
+         * @param args arguments passed to the user functor
+         */
+        template<
+            typename T_Context,
+            typename T_Par0,
+            typename T_Par1
+        >
+        DINLINE void operator( )(
+            T_Context const & ctx,
+            T_Par0 & par0,
+            T_Par1 & par1
+        ) const
+        {
+            //feenableexcept(FE_INVALID | FE_OVERFLOW | FE_DIVBYZERO);
+            float_X const weight0 = par0[ weighting_ ]
+                / particles::TYPICAL_NUM_PARTICLES_PER_MACROPARTICLE;
+            float_X const weight1 = par1[ weighting_ ]
+                / particles::TYPICAL_NUM_PARTICLES_PER_MACROPARTICLE;
+
+            float3_X const labMomentum0 = par0[ momentum_ ] / weight0;
+            float3_X const labMomentum1 = par1[ momentum_ ]  / weight1;
+
+            float_X const mass0 = picongpu::traits::attribute::getMass(
```

Comment:
Note: We should check on the host site where the kernel is called that both species will not be `photon` like particles, because those have the mass zero.

Reply:
@psychocoderHPC how do I do that?